arraykit.c

```diff
@@ -129,6 +130,62 @@ AK_ResolveDTypeIter(PyObject *dtypes)
     return resolved;
 }
 
+// Numpy implementation: https://github.com/numpy/numpy/blob/a14c41264855e44ebd6187d7541b5b8d59bb32cb/numpy/core/src/multiarray/methods.c#L1557
+PyObject*
+AK_ArrayCopy(PyObject *array, PyObject *memo)
```

Comment:
Should we make `array` a `PyArrayObject *` in the signature? We're casting it half the time we use it anyways, and it will help the compiler flag mistakes we make when using this method later.

Also, I would suggest renaming this to `AK_ArrayCopyWithMemo` or something. It's quite a bit more complicated than a blind copy!

Reply:
Agreed on both points. I will name it `AK_ArrayDeepCopy`.